File: js/1921.0_Eliminate_Maximum_Number_of_Monsters.js
/*
132ms 击败 97.50%使用 JavaScript 的用户
52.09MB 击败 87.50%使用 JavaScript 的用户
*/


/**
 * @param {number[]} dist
 * @param {number[]} speed
 * @return {number}
 */
var eliminateMaximum = function(dist, speed) {
    const n = dist.length;
    const arrivalTimes = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
        arrivalTimes[i] = Math.ceil(dist[i] / speed[i]);
    }
    arrivalTimes.sort((a, b) => (a - b));
    for (let i = 0; i < n; i++) {
        if (arrivalTimes[i] <= i) {
            return i;
        }
    }
    return n;
};

File: leetcode-cn/1921.0_Eliminate_Maximum_Number_of_Monsters.py
'''
140ms 击败 47.14%使用 Python3 的用户
29.81MB 击败 38.57%使用 Python3 的用户
'''
class Solution:
    def eliminateMaximum(self, dist: List[int], speed: List[int]) -> int:
        n = len(dist)
        time = [(d + s - 1) // s for d, s in zip(dist, speed)]
        time.sort()
        for i, t in enumerate(time):
            if t <= i:
                return i
        return n 
